pallets/pallet-dip-receiver/src/lib.rs

```diff
@@ -0,0 +1,143 @@
+	#[pallet::getter(fn identity_proofs)]
+	pub(crate) type IdentityProofs<T> =
+		StorageMap<_, Twox64Concat, <T as Config>::Identifier, <T as Config>::ProofDigest>;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Identifier: Parameter + MaxEncodedLen;
+		type ProofLeafKey: Parameter;
+		type ProofLeafValue: Parameter;
+		type ProofDigest: Parameter + MaxEncodedLen;
+		type ProofVerifier: IdentityProofVerifier<
+			ProofDigest = Self::ProofDigest,
+			LeafKey = Self::ProofLeafKey,
+			LeafValue = Self::ProofLeafValue,
+		>;
+		type RuntimeCall: Parameter + Dispatchable<RuntimeOrigin = <Self as Config>::RuntimeOrigin>;
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		type RuntimeOrigin: From<Origin<Self>>
+			+ From<<Self as frame_system::Config>::RuntimeOrigin>
+			+ Into<Result<cumulus_pallet_xcm::Origin, <Self as Config>::RuntimeOrigin>>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		IdentityInfoDeleted(T::Identifier),
+		IdentityInfoUpdated(T::Identifier, T::ProofDigest),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		Dispatch,
+		IdentityNotFound,
+		InvalidProof,
+	}
+
+	// The new origin other pallets can use.
+	#[pallet::origin]
+	pub type Origin<T> = DipOrigin<<T as Config>::Identifier, <T as frame_system::Config>::AccountId>;
+
+	// TODO: Benchmarking
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight(0)]
+		pub fn process_identity_action(
+			origin: OriginFor<T>,
+			action: VersionedIdentityProofAction<T::Identifier, T::ProofDigest>,
+		) -> DispatchResult {
+			ensure_sibling_para(<T as Config>::RuntimeOrigin::from(origin))?;
+
+			let event = match action {
+				VersionedIdentityProofAction::V1(IdentityProofAction::Updated(identifier, proof, _)) => {
+					IdentityProofs::<T>::mutate(&identifier, |entry| *entry = Some(proof.clone()));
+					Event::<T>::IdentityInfoUpdated(identifier, proof)
+				}
+				VersionedIdentityProofAction::V1(IdentityProofAction::Deleted(identifier)) => {
+					IdentityProofs::<T>::remove(&identifier);
+					Event::<T>::IdentityInfoDeleted(identifier)
+				}
+			};
+
+			Self::deposit_event(event);
+
+			Ok(())
+		}
+
+		// TODO: Replace with a SignedExtra.
+		#[pallet::call_index(1)]
+		#[pallet::weight(0)]
+		pub fn dispatch_as(
+			origin: OriginFor<T>,
+			identifier: T::Identifier,
+			proof: VersionedIdentityProofOf<T>,
+			call: Box<<T as Config>::RuntimeCall>,
+		) -> DispatchResult {
+			let submitter = ensure_signed(origin)?;
+			let proof_digest = IdentityProofs::<T>::get(&identifier).ok_or(Error::<T>::IdentityNotFound)?;
+			let _ = T::ProofVerifier::verify_proof_against_digest(proof, proof_digest)
+				.map_err(|_| Error::<T>::InvalidProof)?;
```

Comment:
I don't see the call as part of the proof verification. Shouldn't the proof or something check to make sure that the call is not someone front running and actually requested by the DID owner?

Reply:
Hey, yes, that's in the making! This is the skeleton PR. There's no logic yet in here, but there will be a trait implemented for the `RuntimeCall` type which dictates which calls require the Dip origin, and if so, which key relationship must be used in the signature.
We already use it in our blockchain to define which calls require a DID origin and which don't -> https://github.com/KILTprotocol/kilt-node/blob/e8626944e58585e9ee26ad4e68ddc30413af717f/runtimes/peregrine/src/lib.rs#L996.

Reply:
Also DID signature verification is not yet in there, but it would work similarly to what we have on our chain in the `submit_did_call` extrinsic: https://github.com/KILTprotocol/kilt-node/blob/e8626944e58585e9ee26ad4e68ddc30413af717f/pallets/did/src/lib.rs#L1037.

So the receiver chain would keep track of the last used nonce, so that it's not possible to replay actions nor front-run anything if one action is dependent on another (i.e., the first nonce has to be 1 less than the next one).

Reply:
I see. Cool! Thanks!